app/src/main/java/org/openobservatory/ooniprobe/activity/TextActivity.java

```diff
@@ -39,11 +93,83 @@ public static Intent newIntent(Context context, String text) {
 				Toast.makeText(this, R.string.Toast_CopiedToClipboard, Toast.LENGTH_SHORT).show();
 				return true;
 			default:
 				return super.onOptionsItemSelected(item);
 		}
 	}
+
+	public void showText(){
+		switch (getIntent().getIntExtra(TYPE, 0)) {
+			case TYPE_JSON:
+				//Try to open file, if it doesn't exist dont show Error dialog immediately but try to download the json from internet
+				try {
+					File entryFile = Measurement.getEntryFile(this, measurement.id, measurement.test_name);
+					String json = FileUtils.readFileToString(entryFile, Charset.forName("UTF-8"));
+					json = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create().toJson(new JsonParser().parse(json));
+					text = json;
+					webView.loadDataWithBaseURL(null, json, "text/plain", "UTF-8", null);
+					//webView.loadData(json, "text/plain", "UTF-8");
+				} catch (Exception e) {
+					e.printStackTrace();
+					if (ReachabilityManager.getNetworkType(this).equals(ReachabilityManager.NO_INTERNET)) {
+						new MessageDialogFragment.Builder()
+								.withTitle(getString(R.string.Modal_Error))
+								.withMessage(getString(R.string.Modal_Error_RawDataNoInternet))
+								.build().show(getSupportFragmentManager(), null);
+					}
+					getApiClient().getMeasurement(measurement.report_id, null).enqueue(new GetMeasurementsCallback() {
+						@Override
+						public void onSuccess(ApiMeasurement.Result result) {
+							//Download measurement data locally and displaying into the WebView
+							getOkHttpClient().newCall(new Request.Builder().url(result.measurement_url).build()).enqueue(new GetMeasurementJsonCallback() {
+								@Override
+								public void onSuccess(String json) {
+									text = json;
+									webView.loadData(json, "text/plain", "UTF-8");
+								}
+								@Override
+								public void onError(String msg) {
+									new MessageDialogFragment.Builder()
+											.withTitle(getString(R.string.Modal_Error))
+											.withMessage(msg)
+											.build().show(getSupportFragmentManager(), null);
+								}
+							});
+						}
+						@Override
+						public void onError(String msg) {
+							new MessageDialogFragment.Builder()
+									.withTitle(getString(R.string.Modal_Error))
+									.withMessage(msg)
+									.build().show(getSupportFragmentManager(), null);
+						}
+					});
+				}
+				break;
+			case TYPE_LOG:
+				try {
+					File logFile = Measurement.getLogFile(this, measurement.result.id, measurement.test_name);
+					String log = FileUtils.readFileToString(logFile, Charset.forName("UTF-8"));
+					text = log;
+					webView.loadDataWithBaseURL(null, log, "text/plain", "UTF-8", null);
+				} catch (Exception e) {
+					e.printStackTrace();
+					new MessageDialogFragment.Builder()
+							.withTitle(getString(R.string.Modal_Error_LogNotFound))
+							.build().show(getSupportFragmentManager(), null);
```

Comment:
For robustness, I'd also add a return here

Reply:
From Android Studio : "return' is unnecessary as the last statement in a 'void' method"

Reply:
This is a stupid warning from Android studio, but whatever!